gtfs-validator/main/src/main/java/org/mobilitydata/gtfsvalidator/validator/StopTooFarFromTripShapeValidator.java

```diff
@@ -0,0 +1,131 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.mobilitydata.gtfsvalidator.validator;
+
+import com.google.common.collect.Multimaps;
+import java.util.List;
+import java.util.Map;
+import java.util.HashMap;
+import java.util.ArrayList;
+import org.locationtech.spatial4j.distance.DistanceUtils;
+import org.locationtech.spatial4j.shape.Point;
+import org.locationtech.spatial4j.shape.Shape;
+import org.locationtech.spatial4j.shape.ShapeFactory;
+import org.locationtech.spatial4j.shape.SpatialRelation;
+import org.mobilitydata.gtfsvalidator.annotation.GtfsValidator;
+import org.mobilitydata.gtfsvalidator.annotation.Inject;
+import org.mobilitydata.gtfsvalidator.notice.NoticeContainer;
+import org.mobilitydata.gtfsvalidator.table.GtfsLocationType;
+import org.mobilitydata.gtfsvalidator.table.GtfsShape;
+import org.mobilitydata.gtfsvalidator.table.GtfsShapeTableContainer;
+import org.mobilitydata.gtfsvalidator.table.GtfsStop;
+import org.mobilitydata.gtfsvalidator.table.GtfsStopTableContainer;
+import org.mobilitydata.gtfsvalidator.table.GtfsStopTime;
+import org.mobilitydata.gtfsvalidator.table.GtfsStopTimeTableContainer;
+import org.mobilitydata.gtfsvalidator.table.GtfsTrip;
+import org.mobilitydata.gtfsvalidator.table.GtfsTripTableContainer;
+import org.mobilitydata.gtfsvalidator.notice.StopTooFarFromTripShapeNotice;
+
+import static org.locationtech.spatial4j.context.SpatialContext.GEO;
+
+/**
+ * Validates that a stop is within a distance threshold for a trip shape.
+ * <p>
+ * Generated notices:
+ * * StopTooFarFromTripShapeNotice
+ */
+@GtfsValidator
+public class StopTooFarFromTripShapeValidator extends FileValidator {
+    @Inject GtfsStopTimeTableContainer stopTimeTable;
+
+    @Inject GtfsTripTableContainer tripTable;
+
+    @Inject GtfsShapeTableContainer shapeTable;
+
+    @Inject GtfsStopTableContainer stopTable;
+
+    // Spatial operation buffer values
+    public static final double TRIP_BUFFER_METERS = 100;
+    private static final double TRIP_BUFFER_DEGREES = DistanceUtils.KM_TO_DEG * (TRIP_BUFFER_METERS / 1000.0d);
+
+    @Override
+    public void validate(NoticeContainer noticeContainer) {
+        // Cache for previously tested shape_id and stop_id pairs - no need to test them more than once.
+        final Map<String, List<String>> testedCache = new HashMap<>();
+        // Go through the pair of tripId and the corresponding stop time list one by one.
+        for (Map.Entry<String, List<GtfsStopTime>> tripIdStopTimeListEntry : Multimaps.asMap(stopTimeTable.byTripIdMap()).entrySet()) {
+            final String tripId = tripIdStopTimeListEntry.getKey();
+            final List<GtfsStopTime> stopTimeList = tripIdStopTimeListEntry.getValue();
+            final GtfsTrip trip = tripTable.byTripId(tripId);
+            if (trip == null || stopTimeList == null || stopTimeList.isEmpty() || trip.shapeId() == null) {
```

Comment:
Can stopTimeList be null or empty here? I don't know, but it's possible that they can never be null or never be empty (or both) here because for a tripId to appear in the stopTimeTable, it must be associated with a stopTime.

If this is the case, I think it would be better to avoid checking for null / empty, because a reader might assume that it could be null / empty.

Reply:
If the relationship among tables are complete and no empty table, then yes, there are some null or empty conditions that do not need to be checked for the rule. Can we assume that (and leave these conditions checked in other specific rules)?

Reply:
Yes, the null and empty checks for stopTimeList are unnecessary. I have deleted them.